File: src/descentralized_controller/src/identifier.cpp
#include <ros/ros.h>
#include <std_msgs/String.h>

#include <stdio.h>
#include <sstream>
#include <cmath>
#include <math.h>
#include <Eigen/Dense>

#include <whole_body_state_msgs/JointState.h>
#include <whole_body_state_msgs/WholeBodyTrajectory.h>
#include <whole_body_state_msgs/WholeBodyState.h>
#include <whole_body_state_msgs/RhonnState.h>

#include "Rhonn.h" // Include the Rhonn header
#include "Efk.h"   // Include the Efk header

//Node class
class MyNode{
  public:
    MyNode(ros::NodeHandle& nh) : loop_rate(100) {
      // Initialize ROS node handle, subscriber, and publisher
      nh_ = nh;
      subscriber = nh_.subscribe("/robot_states", 1, &MyNode::subscriberCallback, this);
      publisher = nh_.advertise<whole_body_state_msgs::WholeBodyState>("/reemc/efforts", 1);

      // Init RHONN values (inputs and weights) C ares inputs and W weights
      Eigen::Matrix<float, 2, 1> C1;
      C1 << 1,
            1;
      Eigen::Matrix<float, 3, 1> C2;
      C2 << 1,
            1,
            1;
      // Eigen::Matrix<float, 2, 1> W1 = C1;
      // Eigen::Matrix<float, 3, 1> W2 = C2;
      Eigen::Matrix<float, 2, 1> W1;
/*       W1 << 3.66483,
            -1.83241; */
      W1 << 0,
            0;
      Eigen::Matrix<float, 3, 1> W2;
/*       W2 << 6.64743,
            -13.3101,
            6.64774; */
      W2 << 0,
            0,
            0;
      position = 0;
      position_y1 = 0;
      position_y = 0;
      position_1 = 0;
      velocity = 0;
      velocity_y1 = 0;
      velocity_y = 0;
      velocity_1 = 0;

      // Init values of EFK trainning (P, Q and R)
/*       Eigen::Matrix<float, 1, 1> R1 = 1e-8*Eigen::Matrix<float, 1, 1>::Identity();
      Eigen::Matrix<float, 2, 2> Q1 = 1e-7*Eigen::Matrix<float, 2, 2>::Identity();
      Eigen::Matrix<float, 2, 2> P1 = 1e-10*Eigen::Matrix<float, 2, 2>::Identity();

      Eigen::Matrix<float, 1, 1> R2 = 1e-8*Eigen::Matrix<float, 1, 1>::Identity();
      Eigen::Matrix<float, 3, 3> Q2 = 1e-7*Eigen::Matrix<float, 3, 3>::Identity();
      Eigen::Matrix<float, 3, 3> P2 = 1e-10*Eigen::Matrix<float, 3, 3>::Identity(); */

      Eigen::Matrix<float, 1, 1> R1 = 1e8*Eigen::Matrix<float, 1, 1>::Identity();
      Eigen::Matrix<float, 2, 2> Q1 = 1e7*Eigen::Matrix<float, 2, 2>::Identity();
      Eigen::Matrix<float, 2, 2> P1 = 1e10*Eigen::Matrix<float, 2, 2>::Identity();

      Eigen::Matrix<float, 1, 1> R2 = 1e3*Eigen::Matrix<float, 1, 1>::Identity();
      Eigen::Matrix<float, 3, 3> Q2 = 1e3*Eigen::Matrix<float, 3, 3>::Identity();
      Eigen::Matrix<float, 3, 3> P2 = 1e10*Eigen::Matrix<float, 3, 3>::Identity();

      // Create object Rhonn
      neuron_1 = Rhonn(C1,W1,1);
      neuron_2 = Rhonn(C2,W2,2);

      // Create an object of Efk and update the value of w_weight (HERE CREATE CLASS EFK)
      efk_object_1 = Efk(R1,P1,Q1);
      efk_object_2 = Efk(R2,P2,Q2);
    }

    void subscriberCallback(const whole_body_state_msgs::WholeBodyState& msg) {
      // Create a message to publish
      whole_body_state_msgs::WholeBodyState position_msg;
      whole_body_state_msgs::JointState joint_estimation;
      whole_body_state_msgs::RhonnState rhonn_estimation;

      //for (size_t i = 0; i < 30; i++)
      //{
            // Process the incoming message
            //arm_right_1 is the 21 of 30 joints
            name = msg.joints[21].name;
            position = (msg.joints[21].position*180)/M_PI;
            velocity = (msg.joints[21].velocity*180)/M_PI;

            position_y = position_y1*0.99+position_1*0.00995;
            velocity_y = velocity_y1*0.99+velocity_1*0.00995;
            position_1 = position;
            velocity_1 = velocity;
            position_y1 = position_y;
            velocity_y1 = velocity_y;

            //Change of variable for use those names
            position = position_y;
            velocity = velocity_y;

            // Create a message to publish
            std::cout << "Neurona 1" << std::endl;
            observer_1_value = neuron_1.observer_state(position,velocity);
            neuron_1_value = neuron_1.prediction_state(position, velocity);
            //rhonn_estimation.error_w1 = efk_object_1.error_estimation(neuron_1_value,position,velocity);
            error_1 = efk_object_1.error_estimation(neuron_1_value,position,velocity,1);
            efk_object_1.calculate_new_weights(neuron_1);

            std::cout << "Neurona 2" << std::endl;
            neuron_2.fx0_value(neuron_1);
            observer_2_value = neuron_2.observer_state(position,velocity);
            neuron_2_value = neuron_2.prediction_state(position, observer_2_value);
            //rhonn_estimation.error_w2 = efk_object_2.error_estimation(neuron_2_value,position,velocity);
            error_2 = efk_object_2.error_estimation(neuron_2_value,position,velocity,2);
            efk_object_2.calculate_new_weights(neuron_2);
            neuron_2.control_law(neuron_1, position, velocity);

            //Construction of Message
            rhonn_estimation.name = name;
            rhonn_estimation.position = neuron_1_value;
            rhonn_estimation.velocity = neuron_2_value;
            rhonn_estimation.error_w1 = error_1;
            rhonn_estimation.error_w2 = error_2;
            rhonn_estimation.obs_position = observer_1_value;
            rhonn_estimation.obs_velocity = observer_2_value;

/*             std::cout << "Building message" << std::endl;
            std::cout << "Valor X1_k+1 = " << neuron_1_value << std::endl;
            std::cout << "Valor X2_k+1 = " << neuron_2_value << std::endl; */

            joint_estimation.name = name;
            joint_estimation.position = position;
            joint_estimation.velocity = velocity;
            joint_estimation.effort = neuron_2.getControlLaw();

            //Has to be JointState (joints) NOT Rhonn (rhonn)
            position_msg.joints.push_back(joint_estimation);
            position_msg.rhonn.push_back(rhonn_estimation);
      //}
      
      // Publish the message
      position_msg.header.stamp = ros::Time::now();
      publisher.publish(position_msg);
    }

    void run() {
      // Main loop
      while (ros::ok()) {
          // Call subscriber callback explicitly
          ros::spinOnce();

          // Sleep to maintain the loop rate
          loop_rate.sleep();
      }
    }

    private:
    ros::NodeHandle nh_;
    ros::Subscriber subscriber;
    ros::Publisher publisher;
    ros::Rate loop_rate;
    std::string name;
    float position;
    float position_y1;
    float position_y;
    float position_1;
    float velocity;
    float velocity_y1;
    float velocity_y;
    float velocity_1;
    float neuron_1_value;
    float observer_1_value;
    float neuron_2_value;
    float observer_2_value;
    float error_1;
    float error_2;
    Rhonn neuron_1;
    Rhonn neuron_2;
    Efk efk_object_1;
    Efk efk_object_2;
};

//Main Code Start
int main(int argc, char** argv) {
    // Initialize the ROS node
    ros::init(argc, argv, "identifier");

    //Create a ROS node handle
    ros::NodeHandle nh;

    // Create an instance of MyNode
    MyNode myNode(nh);

    // Run the node
    myNode.run();

    return 0;
}

File: src/descentralized_controller/scripts/identifier.py
#!/usr/bin/env python

import rospy
import numpy as np
import time
import csv
import matplotlib.pyplot as plt
from std_msgs.msg import String
from whole_body_state_msgs.msg import JointState, WholeBodyTrajectory, WholeBodyState, RhonnState
from rhonn import Rhonn
from efk import Efk
from controller import Controller

# Initialize RHONN values (inputs and weights)
C1 = np.array([1, 1], dtype=float).reshape((2, 1))
C2 = np.array([1, 1], dtype=float).reshape((2, 1))

W1 = np.array([0, 0], dtype=float).reshape((2, 1))
W2 = np.array([0, 0], dtype=float).reshape((2, 1))

# Initialize values of EFK training (P, Q, and R)
R1 = 1e5 * np.identity(1, dtype=float)
Q1 = 1e5 * np.identity(2, dtype=float)
P1 = 1e10 * np.identity(2, dtype=float)

R2 = 1e6 * np.identity(1, dtype=float)
Q2 = 1e8 * np.identity(2, dtype=float)
P2 = 1e15 * np.identity(2, dtype=float)


# Create object Rhonn
neuron_1 = Rhonn(C1, W1, 1)
neuron_2 = Rhonn(C2, W2, 2)

# Create object EFK
efk_object_1 = Efk(R1,P1,Q1)
efk_object_2 = Efk(R2,P2,Q2)

#Create object Controller
controller_object = Controller(neuron_1,neuron_2)

# Create variables for message
position_msg = WholeBodyState()
joint_estimation = JointState()
rhonn_estimation = RhonnState()

# Create simple variables
position_1 = 0.0
velocity_1 = 0.0
position_y1 = 0.0
velocity_y1 = 0.0

# Initialize global variables
position = 0.0
velocity = 0.0
name = ""

# Internal Clock Simulation
clock_sim = 0.0
dt = 0.00001
dt_controller = 0.01
time_end = 15.0

def sensor_callback(data):
    global position_msg, joint_estimation, rhonn_estimation, name, position, velocity, position_1, velocity_1, position_y1, velocity_y1

    name = data.joints[21].name
    position = data.joints[21].position
    velocity = data.joints[21].velocity

    # Convert from radian to degrees
    position = (position*180)/np.pi
    velocity = (velocity*180)/np.pi

    # Filter first degree
    position_y = position_y1*0.99+position_1*0.00995
    velocity_y = velocity_y1*0.99+velocity_1*0.00995
    position_1 = position
    velocity_1 = velocity
    position_y1 = position_y
    velocity_y1 = velocity_y

    # Change of variable for use those names
    position = position_y
    velocity = velocity_y

def processing_data(event):
    global position_msg, joint_estimation, rhonn_estimation, name, position, velocity, position_1, velocity_1, position_y1, velocity_y1, clock_sim
    #print("Neurona 1")
    observer_1_value = neuron_1.observer_state(position, velocity)
    neuron_1_value = neuron_1.prediction_state(position,velocity)
    error_1 = efk_object_1.error_estimation(neuron_1_value,position,velocity,1)
    efk_object_1.calculate_new_weights(neuron_1)

    #print("Neuron 2")
    observer_2_value = neuron_2.observer_state(position, velocity)
    neuron_2_value = neuron_2.prediction_state(position,velocity)
    error_2 = efk_object_2.error_estimation(neuron_2_value,position,velocity,2)
    efk_object_2.calculate_new_weights(neuron_2)
    neuron_2.get_control_law(controller_object) #Its just to get u, not calculate u

    

    # Construction of Message
    rhonn_estimation.name = name
    rhonn_estimation.position = neuron_1_value
    rhonn_estimation.velocity = neuron_2_value
    rhonn_estimation.error_w1 = error_1
    rhonn_estimation.error_w2 = error_2
    rhonn_estimation.obs_position = observer_1_value
    rhonn_estimation.obs_velocity = observer_2_value
    rhonn_estimation.reference = controller_object.tetha

    joint_estimation.name = name
    joint_estimation.position = position
    joint_estimation.velocity = velocity
    joint_estimation.effort = neuron_2.u

def control_law_calculation(event):
    controller_object.control_law(position,velocity, clock_sim) #Here is calculated the control law

def save_to_csv(data):
        with open('vector_data.csv', 'wb') as csvfile:
            writer = csv.writer(csvfile)
            writer.writerow(['Time', 'position', 'velocity', 'rhonn_x0', 'rhonn_x1', 'error_x0', 'error_x1', 'set_point', 'effort', 'error_tracking_x0', 'error_tracking_x1', 'surface', 'w11', 'w12', 'w21', 'w22'])  # Assuming 3D vectors
            writer.writerows(data)
        rospy.loginfo('Data saved to vector_data.csv')

def ending_node(pub):
    rospy.loginfo('Ending Program, home position and effort cero')
    global position_msg, joint_estimation, rhonn_estimation, name, position, velocity
    joint_estimation.name = name
    joint_estimation.position = position
    joint_estimation.velocity = velocity
    joint_estimation.effort = 0.5
    position_msg.joints.append(joint_estimation)
    pub.publish(position_msg)

def talker():
    global position_msg, joint_estimation, rhonn_estimation, name, position, velocity, clock_sim
    rospy.init_node('identifier', anonymous=True)
    pub = rospy.Publisher('/reemc/efforts', WholeBodyState, queue_size=1)
    rate = rospy.Rate(1e2) # 10hz
    rospy.Subscriber('/robot_states', WholeBodyState, sensor_callback, queue_size=1)
    rospy.Timer(rospy.Duration(dt), processing_data)
    rospy.Timer(rospy.Duration(dt_controller), control_law_calculation)
    data = []

    #while not rospy.is_shutdown():
    while clock_sim < time_end:   #and left_time > 0

        # Clear the messages to avoid accumulation
        position_msg.joints = []
        position_msg.rhonn = []

        # Has to be JointState (joints) NOT Rhonn (rhonn)
        position_msg.joints.append(joint_estimation)
        position_msg.rhonn.append(rhonn_estimation)

        # Append position and velocity to data list
        data.append([rospy.get_time(), joint_estimation.position, joint_estimation.velocity, rhonn_estimation.position, rhonn_estimation.velocity, rhonn_estimation.error_w1, rhonn_estimation.error_w2, controller_object.tetha, joint_estimation.effort, controller_object.error_x0, controller_object.error_x1, controller_object.s, neuron_1.w_weight[0,0], neuron_1.w_weight[1,0], neuron_2.w_weight[0,0], neuron_2.w_weight[1,0]])

        # Publish the message
        position_msg.header.stamp = rospy.Time.now()
        position_msg.time = rospy.get_time()
        pub.publish(position_msg)
        rate.sleep()
        clock_sim = clock_sim + dt_controller

    save_to_csv(data)
    print("Final Time clock_sim:{}".format(clock_sim))
    ending_node(pub)

if __name__ == '__main__':
    try:
        talker()
    except rospy.ROSInterruptException:
        #ending_node()
        pass
